doc/bakugo_spec.md

```diff
@@ -0,0 +1,1111 @@
+
+A function type denotes the set of all functions with the same parameter and result types. The value of an uninitialized variable of function type is nil.
+
+```
+Signature      = Parameters [ Result ] .
+Result         = Parameters | Type .
+Parameters     = "(" [ ParameterList [ "," ] ] ")" .
+ParameterList  = ParameterDecl { "," ParameterDecl } .
+ParameterDecl  = [ IdentifierList ] Type .
+```
+
+Within a list of parameters or results, the names (IdentifierList) must either all be present or all be absent. If present, each name stands for one item 
+(parameter or result) of the specified type and all non-_blank_ names in the signature must be unique. If absent, each type stands for one item of that type. 
+Parameter and result lists are always parenthesized except that if there is exactly one unnamed result it may be written as an unparenthesized type.
+
+```
+func()
+func(x int) int
+func(a, _ int, z bool) bool
+func(a, b int, z bool) (bool)
+func(prefix bool, values ...int) (success bool)
+```
+
+# Properties of types and values
+
+## Underlying types
+Each type T has an _underlying type_: If T is one of the predeclared boolean, numeric, or string types,
+or a type literal, the corresponding underlying type is T itself. Otherwise, T's underlying type is the
+underlying type of the type to which T refers in its declaration.
+```
+type (
+	A1 = string
+	A2 = A1
+)
+
+type (
+	B1 string
+	B2 B1
+)
+```
+## Core types
+Each non-interface type T has a _core type_, which is the same as the _underlying type_ of T.
+
+## Type identity
+Two types are either _identical_ or _different_.
+
+A _named type_ is always different from any other type. Otherwise, two types are identical if their
+_underlying_ type literals are structurally equivalent; that is, they have the same literal structure and
+corresponding components have identical types.
+```
+type (
+	A0 = float32
+	A1 = A0
+	A2 = bool
+	A3 = int
+
+	B0 A0
+	B1 float32
+
+	C0 = B0
+)
+```
+these types are identical:
+```
+A0, A1, and float32
+A2 and bool
+A3 and int
+
+B0 and C0
+int and int
+```
+
+B0 and B1 are different because they are new types created by distinct type definitions.
+
+## Assignability
+A value x of type V is _assignable_ to a _variable_ of type T ("x is assignable to T") if one of the
+following conditions applies:
+
+- V and T are identical.
+- V and T have identical _underlying types_ and atleast one of V or T is not a _named type_.
+- x is an untyped _constant representable_ by a value of type T.
+
+## Representability
+A _constant_ x is _representable_ by a value of tye T, if one of the following conditions applies:
+
+- x is in the set of value _determined_ by T 
+- T is a floating-point type and x can be rounded to T's precision without overflow. Rounding uses IEEE
+754 round-to-even rules but with an IEEE negative zero further simplified to an unsigned zero. Note that
+constant values never result in an IEEE negative zero, NaN, or infinity.
```

Comment:
We aren't supporting floats yet in the spec, so this doesn't seem necessary

Reply:
Right. I will update that too.